plugins/module_utils/redfish_utils.py

```diff
@@ -3394,3 +3394,85 @@ def get_hpe_fan_percent_min(self):
             fan_percent_min_config = hpe.get('FanPercentMinimum')
         result["fan_percent_min"] = fan_percent_min_config
         return result
+
+    def create_volume(self, volume_details, storage_subsystem_id):
+        # Find the Storage resource from the requested ComputerSystem resource
+        response = self.get_request(self.root_uri + self.systems_uri)
+        if response['ret'] is False:
+            return response
+        data = response['data']
+        storage_uri = data.get('Storage', {}).get('@odata.id')
+        if storage_uri is None:
+            return {'ret': False, 'msg': 'Storage resource not found'}
+
+        # Get Storage Collection
+        response = self.get_request(self.root_uri + storage_uri)
+        if response['ret'] is False:
+            return response
+        data = response['data']
+
+        # Collect Storage Subsystems
+        self.storage_subsystems_uris = [i['@odata.id'] for i in response['data'].get('Members', [])]
+        if not self.storage_subsystems_uris:
+            return {
+                'ret': False,
+                'msg': "StorageCollection's Members array is either empty or missing"}
+
+        # Matching Storage Subsystem ID with user input
+        self.storage_subsystem_uri = ""
+        for storage_subsystem_uri in self.storage_subsystems_uris:
+            if storage_subsystem_uri.split("/")[-2] == storage_subsystem_id:
+                self.storage_subsystem_uri = storage_subsystem_uri
+
+        if not self.storage_subsystem_uri:
+            return {
+                'ret': False,
+                'msg': "Provided Storage Subsystem ID %s does not exist on the server" % storage_subsystem_id}
+
+        # Validate input parameters
+        required_parameters = ['RAIDType', 'Drives']
```

Comment:
I think we should add CapacityBytes to the required list; I can't think of a practical reason as to why we'd let the Redfish service decide the capacity on behalf of the user.

Reply:
Since CapacityBytes parameter is not needed from the API side, should we be restricting it from a code perspective. Also, since the drive link is a mandatory parameter from the API, users have to check the capacity before providing the drives as input.

Reply:
Technically none of these parameters are needed from the API side. Redfish has no required properties in the standard when a user is trying to create a volume. If you look at the Volume schema (https://redfish.dmtf.org/schemas/swordfish/v1/Volume_v1.xml), there are no properties marked as "Redfish.RequiredOnCreate".

A request like this is perfectly acceptable per the specification:

```
POST /redfish/v1/Systems/1/Storage/1/Volumes
Content-Type: application/json

{}
```

With an empty request like this, the client is deferring everything to the service (including the drive selection, RAID type, and size of the volume).

However, is it practical from an Ansible perspective to let a user attempt to create a new Volume without specifying the size? I'm not sure I'm willing to make that leap yet (maybe @felixfontein has an opinion about this).

Reply:
I'm curious what use-cases are there for creating a volume without a specified size. If not specifying the size would mean "use the maximum available" that would make sense to me, but if not specifying it means "pick something for me, I don't really care" (which as far as I understand is the case here?) and there's no other option which implicitly defines the size (like a boolean flag "use the maximum available"), then it feels strange.

One thing to remember is that it's easier to remove restrictions (like initially require the size, and later drop that restriction when someone provides a good argument for it) than adding them (which is usually a breaking change).

Reply:
I understand that your API allows for this; the standard definition technically allows for a completely empty request.

However, from an Ansible perspective, it's not clear what the use case for not specifying the volume size would be. This is not about what the API allows, but rather is it realistic to let an Ansible user create a volume without specifying its size. I can't think of a scenario where a user would not want to specify the size of the volume.

Reply:
Added CapacityBytes as a mandatory parameter.